src/rust-crypto/OutgoingRequestsManager.ts

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2024 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { OlmMachine } from "@matrix-org/matrix-sdk-crypto-wasm";
+
+import { OutgoingRequest, OutgoingRequestProcessor } from "./OutgoingRequestProcessor";
+import { Logger } from "../logger";
+
+/**
+ * OutgoingRequestsManager: responsible for processing outgoing requests from the OlmMachine.
+ * Ensure that only one loop is going on at once, and that the requests are processed in order.
+ */
+export class OutgoingRequestsManager {
+    /** whether {@link stop} has been called */
+    private stopped = false;
+
+    /** whether a task is currently running */
+    private isTaskRunning = false;
+
+    /** queue of requests to be processed once the current loop is finished */
+    private requestQueue: (() => void)[] = [];
+
+    public constructor(
+        private readonly logger: Logger,
+        private readonly olmMachine: OlmMachine,
+        public readonly outgoingRequestProcessor: OutgoingRequestProcessor,
+    ) {}
+
+    /**
+     * Shut down as soon as possible the current loop of outgoing requests processing.
+     */
+    public stop(): void {
+        this.stopped = true;
+    }
+
+    /**
+     * Process the outgoing requests from the OlmMachine.
+     *
+     * This should be called at the end of each sync, to process any requests that have been queued.
+     * In some cases if outgoing requests need to be sent immediately, this can be called directly.
+     *
+     * There is only one request running at once, and the others are queued.
+     * If a request is currently running the queued request will only trigger an additional run.
+     */
+    public async doProcessOutgoingRequests(): Promise<void> {
+        if (this.isTaskRunning) {
+            // If the task is running, add the request to the queue and wait for completion.
+            // This ensures that the requests are processed only once at a time.
+            await new Promise<void>((resolve) => {
+                this.requestQueue.push(resolve);
+            });
+        } else {
+            await this.executeTask();
+        }
+    }
```

Comment:
I don't think we need a whole queue here; a single deferred will do.

For example, we could replace `requestQueue` with `nextLoopDeferred: undefined | IDeferred<void>`. 


```suggestion
    public doProcessOutgoingRequests(): Promise<void> {
        if (!this.isTaskRunning) {
            // The loop is not yet running. kick it off.
            await this.executeTask();
            return;
        }
     
        // The loop is already running, but we need to do another loop. Blah blah avoid race.   
        // Ensure that nextLoopDeferred is set, which tells `executeTask` to do another loop.
        if (!this.nextLoopDeferred) {
            this.nextLoopDeferred = defer();
        }
        
        await this.nextLoopDeferred.promise();
    }        
```

... and then in executeTask, replace the access to `requestQueue` with:

```typescript
let d = this.nextLoopDeferred;
this.nextLoopDeferred = undefined;
// ... loop
d.resolve();
```